core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -264,7 +264,11 @@ class LogSegment(val log: FileRecords,
     var lastIndexEntry = 0
     maxTimestampSoFar = RecordBatch.NO_TIMESTAMP
     try {
-      for (batch <- log.batches(maxMessageSize).asScala) {
+      // record batches can grow during compaction (see KAFKA-5316), so we allow some extra space when reading
+      // the batches. Checking the message size is only used as a way to check corruption before checking the CRC
+      val batches = log.batches(maxMessageSize * 2).asScala
```

Comment:
I am wondering if it's really useful to verify the message size during recovery. We already verified the size when the data is appended in the first place. The only possibility for the size to increase is because of cleaning, which legitimate. Perhaps we could just not check the message size here at all? Alternatively, perhaps we can only check the size for non-compacted topics?